Filesystem paths should not be urlencoded

qrazi opened the pull request:
SchemaReader::loadImport() feeds UrlUtils::resolveRelativeUrl() with $node->ownerDocument->documentURI and $node->getAttribute("schemaLocation") to determine the location of the file. These values are urlencoded. That means for instance that spaces will be formatted as "%20". If an URI refers to a location on a filesystem, because the scheme of the URI is "file", or no scheme is given, then the path should be url decoded.

The tests have been updated in this commit with an assertion to cover that. Furtermore tested on both Windows 7 and Ubuntu 14.04 that "%20" in a path leads to failure of resolving it. Assuming a dir "./foo bar" exists, php -r "echo file_exists('foo%20bar')?'true':'false';echo PHP_EOL;" vs php -r "echo file_exists('foo bar')?'true':'false';echo PHP_EOL;".

This solves the same issue that PR #8 tries to solve, but does so as well when there is no scheme and saves using realpath (which does a resolve of the path as well).

goetas commented:
I discovered that on linux filesystems `A%20a` is a valid filename (and different from `A a`)... so this change will break things.

What about applying this change only on "http" protocol?

qrazi commented:
Ah, I didn't think of that case. I don't really see how it would work by applying it only for "http"? The whole issue is pretty specific to filesystem, i.e. "file" scheme?

Maybe the resolving of the relative URL should include a file_exists() of some sort? In case of "file"-scheme?

/ReminderToSelf: add testcase for a directory including %20 in the name.

qrazi commented:
Ok, I did some more looking into this. `$node->ownerDocument->documentURI` in `SchemaReader::loadImport()` returns a urlencoded string. And if that would be a path on the filesystem with the characters %20 in it, e.g. file:///foo%20bar/, than `$node->ownerDocument->documentURI` would return the string `file:///foo%2520bar/`, so the percentage-character from path would be urlencoded.

Therefor I don't think anymore that my proposed fix is correct, it neglects the fact that the path can contain %20 which should not be decoded. And the filename comes from `$node->getAttribute("schemaLocation")` which does not seem to be urlencoded and should therefor not be decoded.

I haven't made a proper fix yet, I'll get to that.... :)

qrazi commented:
I changed the location of the urldecode. I added unittests. I tried to use vfsStream for testing, but couldn't get it working with a space in the path. So just went short with manually adding some files. I used a branch for this. Perhaps have a look and see if this is somewhat better? If yes, I could clean up the tests (have the dirs created in setUp() and removed in tearDown()).

[https://github.com/qrazi/xsd-reader/tree/pr_10_dev](https://github.com/qrazi/xsd-reader/tree/pr_10_dev)

goetas commented:
To me looks good. Just fix some minor codestyle issue (indentation, comments, indentation changes in composer.json ecc...) (Are you using tabs as indentation char or spaces? I use 4 spaces "almost" everywhere)

vfsStream is a good tool, but if it is too complicated to make it work, do not worry,

A new branch is aslo good. When you are satisfied with you changes I will suggest to you to remove the not-PR related staff.

`"mikey179/vfsStream": "~1"` should stay in `require-dev` section.

Once you are done, you can update this PR so I can do a "final" review... (since in Github you can't change a base branch in a PR, it will not be easy...). You have two option to do it.

* you can `git push -f origin pr_10_dev:master` (this will replace you current master branch with pr_10_dev)
* you can open a new PR with the right branch

qrazi commented:
WIth last commits cleaned up coding styles and Travis CI runs again.

goetas commented:
Hi! For me it looks good! 
Just one last thing: do you mind me if i rebase/squash some of your commits? (we have 10 commits and 5 of this are just typo/code style...)

Can you do it?

qrazi commented:
I'll read up on it and try and get it done. :)

qrazi commented:
Rebase and squashing: done... :)

goetas commented:
Thank you very much! Well done!

qrazi commented:
I'm really happy I made my first contribution to an open source project, and learned a lot about collaborating on GitHub and how to use git. Thank you very much for your patience and support!
